File: apps/multimodal-screener/Android/app/src/main/java/ai/zetic/aiberry/ui/ChatBubble.kt
package ai.zetic.aiberry.ui

import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.keyframes
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * A conversation bubble — the app's question (leading, light card) or the user's
 * answer (trailing, deep-green), in the shared VoiceVitals palette. 1:1 with
 * iOS-Aiberry's `ChatBubble`.
 */
@Composable
fun ChatBubble(text: String, isUser: Boolean) {
    Row(modifier = Modifier.fillMaxWidth()) {
        if (isUser) Spacer(Modifier.width(40.dp))
        if (!isUser) {
            BubbleBody(text, isUser)
            Spacer(Modifier.width(40.dp))
        } else {
            Spacer(Modifier.weight(1f))
            BubbleBody(text, isUser)
        }
    }
}

@Composable
private fun BubbleBody(text: String, isUser: Boolean) {
    Text(
        text = text,
        fontSize = 16.sp,
        fontWeight = if (isUser) FontWeight.Medium else FontWeight.Normal,
        color = if (isUser) Color.White else Theme.Ink,
        modifier = Modifier
            .clip(RoundedCornerShape(18.dp))
            .background(if (isUser) Theme.TileInk else Theme.CardAlt)
            .padding(horizontal = 16.dp, vertical = 12.dp),
    )
}

/** Animated "listening…" placeholder shown in the user's bubble while recording. */
@Composable
fun ListeningBubble() {
    Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.End) {
        Row(
            modifier = Modifier
                .clip(RoundedCornerShape(18.dp))
                .background(Theme.TileInk.copy(alpha = 0.9f))
                .padding(horizontal = 18.dp, vertical = 14.dp),
            horizontalArrangement = Arrangement.spacedBy(5.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            val transition = rememberInfiniteTransition(label = "dots")
            repeat(3) { i ->
                val alpha by transition.animateFloat(
                    initialValue = 0.4f,
                    targetValue = 0.4f,
                    animationSpec = infiniteRepeatable(
                        animation = keyframes {
                            durationMillis = 1200
                            0.4f at 0 using LinearEasing
                            1f at (i * 400 + 200) using LinearEasing
                            0.4f at (i * 400 + 400) using LinearEasing
                        },
                        repeatMode = RepeatMode.Restart,
                    ),
                    label = "dot$i",
                )
                Box(
                    modifier = Modifier
                        .size(7.dp)
                        .alpha(alpha)
                        .clip(CircleShape)
                        .background(Color.White),
                )
            }
        }
    }
}

File: apps/multimodal-screener/iOS/Aiberry/UI/ChatBubble.swift
import SwiftUI

/// A conversation bubble — the app's question (leading, light card) or the user's
/// answer (trailing, deep-green), in the shared VoiceVitals palette.
struct ChatBubble: View {
    enum Role { case prompt, user }
    let role: Role
    let text: String

    var body: some View {
        HStack {
            if role == .user { Spacer(minLength: 40) }
            Text(text)
                .font(role == .prompt ? .body : .body.weight(.medium))
                .foregroundStyle(role == .prompt ? Theme.ink : .white)
                .padding(.horizontal, 16).padding(.vertical, 12)
                .background(role == .prompt ? Theme.cardAlt : Theme.tileInk)
                .clipShape(RoundedRectangle(cornerRadius: 18, style: .continuous))
            if role == .prompt { Spacer(minLength: 40) }
        }
    }
}

/// Animated "listening…" placeholder shown in the user's bubble while recording.
struct ListeningBubble: View {
    @State private var phase = 0
    private let timer = Timer.publish(every: 0.4, on: .main, in: .common).autoconnect()

    var body: some View {
        HStack {
            Spacer(minLength: 40)
            HStack(spacing: 5) {
                ForEach(0..<3) { i in
                    Circle().fill(.white.opacity(phase == i ? 1 : 0.4))
                        .frame(width: 7, height: 7)
                }
            }
            .padding(.horizontal, 18).padding(.vertical, 14)
            .background(Theme.tileInk.opacity(0.9))
            .clipShape(RoundedRectangle(cornerRadius: 18, style: .continuous))
        }
        .onReceive(timer) { _ in phase = (phase + 1) % 3 }
    }
}
